File: kafka-streams-java/7-windowed-leaderboard/src/main/java/com/github/simplesteph/udemy/kafka/streams/internal/Generator.java
package com.github.simplesteph.udemy.kafka.streams.internal;

import com.github.simplesteph.Character;
import com.github.simplesteph.Game;
import com.github.simplesteph.Player;
import com.github.simplesteph.Victory;
import org.jeasy.random.EasyRandom;
import org.jeasy.random.EasyRandomParameters;

import java.time.Instant;
import java.util.*;

import static org.jeasy.random.util.CollectionUtils.randomElementOf;

public abstract class Generator {

    private static final ArrayList<Character> sfCharacters = new ArrayList<>(Arrays.asList(
            new Character("Ken", "US"),
            new Character("Ryu", "Japan"),
            new Character("Geki", "Japan"),
            new Character("Chun-Li", "China"),
            new Character("Akuma", "Japan"),
            new Character("Sakura", "Japan"),
            new Character("Dhalsim", "India"),
            new Character("Blair", "UK")
    ));

    private static final ArrayList<Character> takkenCharacters = new ArrayList<>(Arrays.asList(
            new Character("Jin", "Japan"),
            new Character("Asuka", "Japan"),
            new Character("Emilie", "Monaco"),
            new Character("Kazuya", "Japan")
    ));

    private static final ArrayList<Character> kofCharacters = new ArrayList<>(Arrays.asList(
            new Character("Mai", "Japan"),
            new Character("Ramon", "Mexico"),
            new Character("Nelson", "Brazil"),
            new Character("Vanessa", "France")
    ));

    private static final ArrayList<Character> scCharacters = new ArrayList<>(Arrays.asList(
            new Character("Kilik", "China"),
            new Character("Ivy", "UK"),
            new Character("Siegfried", "HRE"),
            new Character("Nightmare", "X")
    ));

    private static final ArrayList<Character> smCharacters = new ArrayList<>(Arrays.asList(
            new Character("Galford", "US"),
            new Character("Charlotte", "France"),
            new Character("Haohmaru", "Japan"),
            new Character("Ukyo Tachibana", "Japan")
    ));

    private static final HashMap<Game, List<Character>> CHARACTERS_MAP = new HashMap<Game, List<Character>>() {{
        put(Game.KingOfFighters, kofCharacters);
        put(Game.SamuraiShodown, smCharacters);
        put(Game.StreetFighter, sfCharacters);
        put(Game.SoulCalibur, scCharacters);
        put(Game.Takken, takkenCharacters);
    }};

    private static EasyRandomParameters parameters = new EasyRandomParameters()
            .seed(60L)
            .objectPoolSize(30)
            .ignoreRandomizationErrors(true)
            .randomize(Victory.class, () -> {
                Game game = nextGame();
                Character character = nextCharacter(game);
                String termId = Integer.toString(new Random().nextInt(3));
                return new Victory(
                        game,
                        termId,
                        character,
                        Instant.now(),
                        randomElementOf(Arrays.asList(Player.values()))
                );
            });

    private static EasyRandom easyRandom = new EasyRandom(parameters);

    private static Game nextGame() {
        return randomElementOf(Arrays.asList(Game.StreetFighter, Game.Takken, Game.KingOfFighters));
    }

    private static Character nextCharacter(Game game) {
        return randomElementOf(CHARACTERS_MAP.get(game));
    }

    protected static Optional<Victory> nextVictory() {
        return Optional.ofNullable(easyRandom.nextObject(Victory.class));
    }

    protected static int nextDelay() {
        return 1 + easyRandom.nextInt(5);
    }
}

File: kafka-streams-scala/7-windowed-leaderboard-scala/src/main/scala/com/github/simplesteph/udemy/kafka/streams/internal/Generator.scala
package com.github.simplesteph.udemy.kafka.streams.internal

import java.time.Instant

import com.github.simplesteph.{Character, Game, Player, Victory}
import org.scalacheck.Gen

trait Generator {

  private val sfCharacters = Array(
    Character("Ken", "US"),
    Character("Ryu", "Japan"),
    Character("Geki", "Japan"),
    Character("Chun-Li", "China"),
    Character("Akuma", "Japan"),
    Character("Sakura", "Japan"),
    Character("Dhalsim", "India"),
    Character("Blair", "UK")
  )

  private val takkenCharacters = Array(
    Character("Jin", "Japan"),
    Character("Asuka", "Japan"),
    Character("Emilie", "Monaco"),
    Character("Kazuya", "Japan")
  )

  private val kofCharacters = Array(
    Character("Mai", "Japan"),
    Character("Ramon", "Mexico"),
    Character("Nelson", "Brazil"),
    Character("Vanessa", "France")
  )

  private val scCharacters = Array(
    Character("Kilik", "China"),
    Character("Ivy", "UK"),
    Character("Siegfried", "HRE"),
    Character("Nightmare", "X")
  )

  private val smCharacters = Array(
    Character("Galford", "US"),
    Character("Charlotte", "France"),
    Character("Haohmaru", "Japan"),
    Character("Ukyo Tachibana", "Japan")
  )

  private val CharactersMap = Map(
    Game.KingOfFighters -> kofCharacters,
    Game.SamuraiShodown -> smCharacters,
    Game.StreetFighter -> sfCharacters,
    Game.SoulCalibur -> scCharacters,
    Game.Takken -> takkenCharacters
  )

  val victoryGen: Gen[Victory] = for {

    id <- Gen.chooseNum(0, 2)

    game <- Gen.oneOf(Game.values())

    character <- Gen.oneOf(CharactersMap(game))

    player <- Gen.oneOf(Player.values())

  } yield Victory(game, id.toString, character, Instant.now(), player)

  protected def nextVictory: Option[Victory] = victoryGen.sample

  protected def nextDelay: Option[Int] = Gen.chooseNum(1, 3).sample

}
